### PR TITLE
ndk: Mark all enums as `non_exhaustive` and fix `repr` types

### PR DESCRIPTION
**Pull request by MarijnS95**
Depends on #458

We're currently in a painful situation, both when it comes to correctness of the wrapper functions that take or return enum values, as well as convenience and ability to upgrade. This is a summation of issues, and reasons why external crates like `android-activity` are copy-pasting and fixing up our implementation:
- False mapping of unknown values to NDK `_UNKNOWN` constants.  This makes the API lossy as the original unknown value does not have the **KNOWN** value that is assigned to the `_UNKNOWN` constant (typically `0`);
- Lacking `#[non_exhaustive]` in most places, not allowing us to add new **and missing** constants to `enum`;
- `#[repr(u32)]` that follows the default `u32` type used by `bindgen` when processing an "untyped" enum, rather than considering the type (`i32`, `i8`, etc) that is actually used by every function taking or returning this "integer constant";
- Inconsistent use of either:
  - Panicking on unknown values;
  - Wrapping unknown values in a custom `Err` type (i.e. `AudioError::UnknownValue` or `TryFromPrimitiveError`);
  - Assigning unknown values to an `Unknown = ffi::XXX_UNKNOWN` variant;
  - Sometimes returning them in a typed `Unknown(xxx)` variant.

To solve this we have to make the following changes to every `pub enum` with custom discriminants in the NDK crate:

- Drop all builtin type conversions and update `repr(u32)` to use the common type used by _every_ function that takes or returns this enum value. Conversions always happen via `.into()` from `IntoPrimitive` (needed to automatically process `__Unknown`);
- Annotate all those `enum`s with `#[non_exhaustive]`.  This requires callers to have a `x => todo!("New constant {x:?}")` or similar handling in a `match` block.  It will neatly print the value for `__Unknown` added below;
- Replace `TryFromPrimitive` with `FromPrimitive`;
- Have a catch-all `__Unknown` variant, that is `#[doc(hidden)]`.  This:
  - Saves us from having a `TryIntoPrimitiveError` or custom `Err` type;
  - Allows the caller to pass in custom/unknown values;
  - Entices contributions/additions in the crate by having the type `#[doc(hidden)]`, as a lot of enums are currently lagging behind their upstream representation;
- Add `IntoPrimitive` to all enums that were using manual `match`es. `FromPrimitive` is not derived, and similarly an `__Unknown` variant is not added.  This does not allow custom values to be passed, but it does again entice crate updates when there are new values.

# TODO

- [x] For error types like `AudioResult`/`AudioError`, implement `Error` on them immediately like on `MediaError`.  This is possible now that the custom `UnknownValue` variant is gone.
- [x] Apply the same changes to native_window, data_space, and hardware_buffer, for which I have purpose-built documented commits on a separate branch, but might/should as well squash in here. `DataSpace` is fairly recent and I don't remember why I didn't follow the above guidelines right from the get-go.

In a followup PR missing constants will be added to various `enum`s.

### CODE REVIEW

File: ndk/src/data_space.rs [hunk 1]
```diff
@@ -5,13 +5,13 @@
...
 /// Describes how to interpret colors.
 ///
 /// <https://developer.android.com/ndk/reference/group/a-data-space#group___a_data_space_1ga2759ad19cae46646cc5f7002758c4a1c>
-#[repr(u32)]
-#[derive(Clone, Copy, Hash, PartialEq, Eq, TryFromPrimitive, IntoPrimitive)]
+#[repr(i32)]
+#[derive(Clone, Copy, Hash, PartialEq, Eq, FromPrimitive, IntoPrimitive)]
 #[doc(alias = "ADataSpace")]
 #[non_exhaustive]
 pub enum DataSpace {
```
Comment (MarijnS95) [inline ndk/src/data_space.rs:17]: Note that the native type here will change to `i32`:

https://android-review.googlesource.com/c/platform/frameworks/native/+/2902242

File: ndk/src/native_window.rs [hunk 2]
```diff
@@ -409,39 +386,32 @@ bitflags::bitflags! {
     /// of those basic transforms.
     #[derive(Clone, Copy, Debug, Hash, PartialEq, Eq)]
     #[doc(alias = "ANativeWindowTransform")]
-    pub struct NativeWindowTransform : u32 {
+    pub struct NativeWindowTransform : i32 {
         #[doc(alias = "ANATIVEWINDOW_TRANSFORM_IDENTITY")]
-        const TRANSFORM_IDENTITY = ffi::ANativeWindowTransform::ANATIVEWINDOW_TRANSFORM_IDENTITY.0;
+        const IDENTITY = ffi::ANativeWindowTransform::ANATIVEWINDOW_TRANSFORM_IDENTITY.0 as i32;
```
Comment (rib) [inline ndk/src/native_window.rs:391]: minor, unrelated API change here with removing the prefix but this seems good to me

File: ndk/src/native_window.rs [hunk 3]
```diff
@@ -409,39 +386,32 @@ bitflags::bitflags! {
...
-
 /// Compatibility value for [`NativeWindow::set_frame_rate()`]
 #[cfg_attr(
     feature = "api-level-31",
     doc = " and [`NativeWindow::set_frame_rate_with_change_strategy()`]"
 )]
 /// .
 #[cfg(feature = "api-level-30")]
-#[repr(u32)]
+#[repr(i8)]
```
Comment (rib) [inline ndk/src/native_window.rs:414]: Not sure why this is `i8` instead if `i32` - By default I'd generally expect most NDK enums should effectively be represented with the equivalent of a JNI `jint`, since most of them are effectively defined in the Android Java SDK

aah OK, I see this particular enum is passed as an argument to an API that takes an `int8_t` which implies they can't really extend beyond `i8`

**Review by rib — APPROVED** [review]
Nice, this must have been quite a chore to go through and update - kudos!

I made a couple of minor comments, but overall it looks good to me!

**Comment by MarijnS95** [comment]
> Nice, this must have been quite a chore to go through and update - kudos!

Thanks! Besides reviewing the changes here, I hope you've also cross-checked that I didn't miss out on some `enum`s?

Are you okay with the proposed idea of _both_ having `non_exhaustive` and a _hidden_ `__Unknown`?

I'll process the review ASAP and get this in to unblock many other changes, many thanks for the review!

File: ndk/src/native_window.rs (hunk 3 above)
Comment (MarijnS95) [reply ndk/src/native_window.rs:414]: Ah yeah, that's documented in https://github.com/rust-mobile/ndk/pull/459/commits/da1a7b03f2631f5858e51f983018c2f16d1b1527.

I've been having some chats with the Android folks about fixing their type annotations. Since using clang they're looking to switch to an `enum <name>: <type>` representation and replace e.g. `i8` types in function signatures with `enum FrameRateStrategy` etc.

File: ndk/src/native_window.rs (hunk 2 above)
Comment (MarijnS95) [reply ndk/src/native_window.rs:391]: Yeah that was done in a separate commit; it's documented in the changelog but none of the other changes are :sweat_smile:

File: ndk/src/bitmap.rs [hunk 4]
```diff
@@ -57,26 +59,33 @@ fn construct<T>(with_ptr: impl FnOnce(*mut T) -> i32) -> Result<T> {
...
+    A_8 = ffi::AndroidBitmapFormat::ANDROID_BITMAP_FORMAT_A_8.0 as i32,
     #[doc(alias = "ANDROID_BITMAP_FORMAT_RGBA_F16")]
-    RGBA_F16 = ffi::AndroidBitmapFormat::ANDROID_BITMAP_FORMAT_RGBA_F16.0,
+    RGBA_F16 = ffi::AndroidBitmapFormat::ANDROID_BITMAP_FORMAT_RGBA_F16.0 as i32,
     #[doc(alias = "ANDROID_BITMAP_FORMAT_RGBA_1010102")]
-    RGBA_1010102 = ffi::AndroidBitmapFormat::ANDROID_BITMAP_FORMAT_RGBA_1010102.0,
+    RGBA_1010102 = ffi::AndroidBitmapFormat::ANDROID_BITMAP_FORMAT_RGBA_1010102.0 as i32,
+
+    // Use the SUCCESS discriminant, as no-one will be able to call `as i32` and only has access to
+    // the constants via `From` provided by `IntoPrimitive` which reads the contained value.
```
[the remark this answers is not in the document]
Comment (MarijnS95) [reply ndk/src/bitmap.rs:85]: If I recall correctly, the exact reason with one of the enums was that it went like:

```
enum XXX {
   Ok = 0,
   Err1 = -1,
   Err2 = -2,

   __Unknown(..) = (Err2 + 1),
```

This is done implicitly by the compiler, but would fail to compile because of overlap with `Err1`.

[the remark this answers is not in the document]
Comment (MarijnS95) [reply ndk/src/bitmap.rs:85]: It is `BitmapError`:

```rust
error[E0081]: discriminant value `-1` assigned more than once
  --> ndk/src/bitmap.rs:22:1
   |
22 | pub enum BitmapError {
   | ^^^^^^^^^^^^^^^^^^^^
...
26 |     BadParameter = ffi::ANDROID_BITMAP_RESULT_BAD_PARAMETER,
   |                    ---------------------------------------- `-1` assigned here
27 |     #[doc(alias = "ANDROID_BITMAP_RESULT_JNI_EXCEPTION")]
28 |     JniException = ffi::ANDROID_BITMAP_RESULT_JNI_EXCEPTION,
   |     ------------ discriminant for `__Unknown` incremented from this startpoint (`JniException` + 1 variant later => `__Unknown` = -1)
...
34 |     __Unknown(i32),
   |     --------- `-1` assigned here
```

[the remark this answers is not in the document]
Comment (MarijnS95) [reply ndk/src/bitmap.rs:85]: I'll clear out the rest though, should only be necessary in a few cases.

**Comment by rib** [comment]
> I hope you've also cross-checked that I didn't miss out on some `enum`s?

I didn't, but can take a pass to cross reference now...

File: ndk/src/bitmap.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (rib) [reply ndk/src/bitmap.rs:85]: Okey, yep, makes sense

[the remark this answers is not in the document]
Comment (MarijnS95) [reply ndk/src/bitmap.rs:85]: It doesn't really matter though _what_ discriminant we use, `.into()` from `IntoPrimitive` takes care of never returning it and the user is disallowed from using `as <primitive>`.

**Comment by rib** [comment]
Going through cross checking all enums, it could be good go add `non_exhaustive` to `InputEvent` potentially.

At least in `android-activity` I defined `InputEvent` as `non_exhaustive` considering also exposing `Text`/IME events but maybe for the ndk API it's ok to just assume the types of InputQueue events won't get extended?

**Comment by rib** [comment]
> Are you okay with the proposed idea of both having non_exhaustive and a hidden __Unknown?

Yeah I think it makes sense. This is at least what I settled on with android-activity after trying to ponder the trade offs. It's maybe not the perfect solution - ideally the compiler could give us a true `repr(i32)` enum but it seems like a pretty decent trade off for allowing extension of enums via Android updates.

**Review by rib — CHANGES_REQUESTED** [review]
It probably makes sense to add an `__Unknown` variant to `FamilyVariant` I think

**Comment by rib** [comment]
> > I hope you've also cross-checked that I didn't miss out on some `enum`s?
> 
> I didn't, but can take a pass to cross check now...

okey, I've taken a pass through all source for the `ndk` crate to try and find any enums missed and left a few comments above and haven't found anything else.

File: ndk/src/native_window.rs (hunk 2 above)
Comment (MarijnS95) [reply ndk/src/native_window.rs:391]: After some realization, I already did open a PR to separate this change out (hence the separate commit), this PR just depends on that one.

https://github.com/rust-mobile/ndk/pull/460

Mind giving it your approval so that I can get a merge-train going (after addressing all the comments/cleanups going)?

File: ndk/src/bitmap.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (MarijnS95) [reply ndk/src/bitmap.rs:85]: By the way, for `aaudio_result_t` the variants are listed from lowest to highest, (since they have `base=-900` and then use `disconnected=base+1, illegal_argument=base+2` andsoforth). We could skip the discriminant here and have the Rust code generate one that might overlap with an upcoming variant in the NDK, again because `num_enum` should be able to deal with that and the tuple variant prevents calling a plain `as i32`.

```rust
pub const AAUDIO_ERROR_BASE: _bindgen_ty_50 = -900;
pub const AAUDIO_ERROR_DISCONNECTED: _bindgen_ty_50 = -899;
pub const AAUDIO_ERROR_ILLEGAL_ARGUMENT: _bindgen_ty_50 = -898;
pub const AAUDIO_ERROR_INTERNAL: _bindgen_ty_50 = -896;
pub const AAUDIO_ERROR_INVALID_STATE: _bindgen_ty_50 = -895;
pub const AAUDIO_ERROR_INVALID_HANDLE: _bindgen_ty_50 = -892;
pub const AAUDIO_ERROR_UNIMPLEMENTED: _bindgen_ty_50 = -890;
pub const AAUDIO_ERROR_UNAVAILABLE: _bindgen_ty_50 = -889;
pub const AAUDIO_ERROR_NO_FREE_HANDLES: _bindgen_ty_50 = -888;
pub const AAUDIO_ERROR_NO_MEMORY: _bindgen_ty_50 = -887;
pub const AAUDIO_ERROR_NULL: _bindgen_ty_50 = -886;
pub const AAUDIO_ERROR_TIMEOUT: _bindgen_ty_50 = -885;
pub const AAUDIO_ERROR_WOULD_BLOCK: _bindgen_ty_50 = -884;
pub const AAUDIO_ERROR_INVALID_FORMAT: _bindgen_ty_50 = -883;
pub const AAUDIO_ERROR_OUT_OF_RANGE: _bindgen_ty_50 = -882;
pub const AAUDIO_ERROR_NO_SERVICE: _bindgen_ty_50 = -881;
pub const AAUDIO_ERROR_INVALID_RATE: _bindgen_ty_50 = -880;
```

Our `__Unknown` variant implicitly gets discriminant `-879`.

File: ndk/src/configuration.rs [hunk 5]
```diff
@@ -162,20 +162,12 @@ impl Configuration {
 
     /// Returns the keyboard type.
     pub fn keyboard(&self) -> Keyboard {
-        unsafe {
-            (ffi::AConfiguration_getKeyboard(self.ptr.as_ptr()) as u32)
-                .try_into()
-                .unwrap()
-        }
+        unsafe { ffi::AConfiguration_getKeyboard(self.ptr.as_ptr()).into() }
```
[the remark this answers is not in the document]
Comment (MarijnS95) [reply ndk/src/configuration.rs:165]: Yeah, returning `TryFromPrimitiveError` is annoying to the user (while also not being able to safely pass _unknown_ values into the API - even though we want to _discourage_ that), this is just outright impossible for new enum values that will show up.

**Comment by MarijnS95** [comment]
> Going through cross checking all enums, it could be good go add `non_exhaustive` to `InputEvent` potentially.
> 
> At least in `android-activity` I defined `InputEvent` as `non_exhaustive` considering also exposing `Text`/IME events but maybe for the ndk API it's ok to just assume the types of InputQueue events won't get extended?

Good point. I ended up glossing over this but there are definitely more input types, though currently none has a type and specific functions. Let's mark this as `non_exhaustive`.

```
pub const AINPUT_EVENT_TYPE_KEY: _bindgen_ty_19 = 1;
pub const AINPUT_EVENT_TYPE_MOTION: _bindgen_ty_19 = 2;
pub const AINPUT_EVENT_TYPE_FOCUS: _bindgen_ty_19 = 3;
pub const AINPUT_EVENT_TYPE_CAPTURE: _bindgen_ty_19 = 4;
pub const AINPUT_EVENT_TYPE_DRAG: _bindgen_ty_19 = 5;
pub const AINPUT_EVENT_TYPE_TOUCH_MODE: _bindgen_ty_19 = 6;
```

**Comment by MarijnS95** [comment]
> It probably makes sense to add an `__Unknown` variant to `FamilyVariant` I think

Whoops, looks like I made an exception for an input-only enum here, to encourage NDK updates when needing to pass new values. Didn't do that anywhere else I think, so let's add it.

File: ndk/src/audio.rs [hunk 6]
```diff
@@ -434,20 +483,17 @@ impl AudioResult {
     }
 }
 
+// TODO: Delete this and `impl Error for AudioResult`, just like `MediaError`!
```
[the remark this answers is not in the document]
Comment (MarijnS95) [reply ndk/src/audio.rs:486]: I've just tackled it immediately in https://github.com/rust-mobile/ndk/pull/467 instead, thanks for reminding me!

In hindsight it could have been done in this PR as we're already breaking the `AudioError` type anyway.

**Comment by MarijnS95** [comment]
Thanks a lot @rib! I know it's a big ask to go through _all_ enums, it's quite a bit to process. But glad to have it resolved in the end.

Who knows, once I import the new values, if we can just replace the enums in `android-activity` with these again.

File: ndk/src/bitmap.rs [hunk 7]
```diff
@@ -26,10 +26,14 @@ pub enum BitmapError {
     BadParameter = ffi::ANDROID_BITMAP_RESULT_BAD_PARAMETER,
     #[doc(alias = "ANDROID_BITMAP_RESULT_JNI_EXCEPTION")]
     JniException = ffi::ANDROID_BITMAP_RESULT_JNI_EXCEPTION,
-    // Use the OK discriminant, as no-one will be able to call `as i32` and only has access to the
-    // constants via `From` provided by `IntoPrimitive` which reads the contained value.
+
+    // Use the SUCCESS discriminant, as no-one will be able to call `as i32` and only has access to
+    // the constants via `From` provided by `IntoPrimitive` which reads the contained value.
+    // An autogenerated `<previous variant> + 1` discrriminant is normally fine, except that the
```
Comment (rib) [inline ndk/src/bitmap.rs:32]:
```suggestion
    // An autogenerated `<previous variant> + 1` discriminant is normally fine, except that the
```

**Review by rib — APPROVED** [review]
Cool, I think this is all looking good! Really good stuff!

File: ndk/src/bitmap.rs (hunk 7 above)
Comment (MarijnS95) [reply ndk/src/bitmap.rs]: Found the same c-p error in `media_error.rs` :)

**Comment by MarijnS95** [comment]
It seems like I never mentioned it, but another "advantage" (though it is far-fetched) of hiding `__Unknown()` as much as possible is that hopefully users don't start to rely on `PartialEq, Eq, Hash`. That is, if the user constructs an `__Unknown()` with a value that we actually know (or start to define in a followup *nonbreaking patch release*) it won't actually be equal. I.e. "contrary to expected behaviour": `Dataspace::__Unknown(DataSpace::Bt2020 as i32) != DataSpace::Bt2020`.

**Comment by rib** [comment]
> It seems like I never mentioned it, but another "advantage" (though it is far-fetched) of hiding `__Unknown()` as much as possible is that hopefully users don't start to rely on `PartialEq, Eq, Hash`. That is, if the user constructs an `__Unknown()` with a value that we actually know (or start to define in a followup _nonbreaking patch release_) it won't actually be equal. I.e. "contrary to expected behaviour": `Dataspace::__Unknown(DataSpace::Bt2020 as i32) != DataSpace::Bt2020`.

hmm, good point, though I guess ideally `PartialEq/Eq/Hash` should all be unconditionally based on the numeric value of underlying Android SDK constant to avoid this problem.

The benefit of the `__Unknown` variant is that code is able to deal with values from the Android SDK at runtime that it didn't know about at compile time, but then as new variants are added it's possible that downstream crates won't all update to use the new Rust enum variants at the same time.

E.g.an up-to-date crate A could create a `DataSpace::Bt2020` while an older crate B creates a `Dataspace::__Unknown(DataSpace::Bt2020 as i32)` and then if those values are passed to a crate C that compares then ideally they would be considered equal.

**Comment by MarijnS95** [comment]
@rib yes, that is exactly what I meant. Currently it's not based purely on NDK constant values, making this all pretty inconvenient. The `!=` sample above should actually be considered equal. That is what I've been looking at over on `drm-fourcc-rs`.

Additionally though `num_enum` doesn't allow it, we could try to make the contained value/type hidden and private. That way a user can only store a custom unknown value via a `from_i32(i32) -> DataSpace` for example, forcing the the value to be converted to the known variant for the given `ndk` crate version. Likewise, their matches would have to have `x @ DataSpace::__Unknown(_) => x.into_i32()` for example.
